### ISSUE TITLE
Static diagram sizing option

### ISSUE THREAD

**Issue by pangloss** [terrastruct/d2#1080]
Hi, I use d2 on a daily basis for analyzing program graphs for the compiler I'm working on. It works great for me in general but I'd love to convince you to make diagram scaling optional. 

Given the wide variation of both diagram sizes user viewport sizes, I think viewport scaling must rarely be the optimal setting. Most likely only when it happens to produce graphs that are close to the diagram's natural size anyway. I really wish that I could change this behaviour.

My diagrams are sometimes quite large, often with a long sequence of steps rendered either top to bottom or left to right.

In early versions of d2, large graphs were no problem because the graph was not scaled. Everything was a fixed size and I could simply pan around in the browser, and possibly zoom out if I needed to orient myself.

Once viewport scaling came in, things became much more complex for me. In In order for the graph to be legible, I must zoom my browser out as far as possible, reload the page, then zoom in as far as possible. This involves something like 20 steps on the zoom controls. In fact, I had to customize Firefox to allow me a wider range of zoom steps because I needed nearly 100x zoom to read the text on one of the larger diagrams I wanted to look at!

If the diagram has too many elements, it frequently comes out looking like Fig. 1 below.

What I would like is in Fig. 2. To produce that, I opened up the web inspector and made just one edit! By removing the width and height attributes from the svg tag, the diagram magically re-scaled itself to exactly the right size!

It's very inconvenient to mess with the zoom or go in and edit the svg. I really wish I could just provide an `--unscaled` option when I start the d2 watcher on the console...

Thanks!
Darrick

Before:
```html
<svg id="d2-svg" class="d2-2154263955" width="25.638187993899898" height="953.9999999999999" viewBox="-89 -89 2477 57704">
```
Fig. 1 "Diff view of 15 compiler transform steps, rendered by d2 in 10.7s"
![Difficult to read diagram](https://user-images.githubusercontent.com/10115/227060760-b1c5f9f4-f766-4fb4-bff1-179e94ac7351.jpg)

After
```html
<svg id="d2-svg" class="d2-2154263955" viewBox="-89 -89 2477 57704">
``` 
Fig. 2:
<img width="1709" alt="image" src="https://user-images.githubusercontent.com/10115/227063888-8ea5776b-89d5-4faf-b3e0-3bc18348d66f.png">

**Comment by alixander** [comment]
thanks for the feedback @pangloss , we'll have this for the next release: https://github.com/terrastruct/d2/pull/721#issuecomment-1445154505

**Comment by bo-ku-ra** [comment]
what i really want are
* fit only the width
* fit only the length
* and this

**Comment by alixander** [comment]
@bo-ku-ra what about if direction is left/right, fit height, if direction is up/down, fit width?

and this.

**Comment by bo-ku-ra** [comment]
@alixander 
that may be the tendency, but it's a matter of preference.
fitting it on one side can make it appear larger.
it's a choice of whether to scroll vertically or horizontally for the viewer to easily understand it, I believe.

**Comment by pangloss** [comment]
I experimented a bit more with a large diagram today. I misunderstood how the browser behaves when the width and height attributes are removed from the svg tag. 

To get a fixed scale version of the diagram, the only change I needed to make was to add `height` and `width` attributes to the outer svg tag. That worked but the diagram was very zoomed in. It was much better when I divided both of those values by 2. Here is the before and after with the first few opening tags of the file:

Before:
```html
<?xml version="1.0" encoding="utf-8"?>
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" d2Version="0.2.6" 
  preserveAspectRatio="xMinYMin meet" viewBox="0 0 7681 3045">
<svg id="d2-svg" class="d2-1358957321" width="7681" height="3045" viewBox="-89 -89 7681 3045">
```
<img width="730" alt="image" src="https://user-images.githubusercontent.com/10115/227421736-278137d8-c4c7-46fa-be4b-6bd5dc61b9f2.png">

After:
```html
<?xml version="1.0" encoding="utf-8"?>
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" d2Version="0.2.6" 
  preserveAspectRatio="xMinYMin meet" viewBox="0 0 7681 3045" 
  width="3840" height="1522">
<svg id="d2-svg" class="d2-1358957321" width="7681" height="3045" viewBox="-89 -89 7681 3045">
```
<img width="735" alt="image" src="https://user-images.githubusercontent.com/10115/227421955-a7f16ee9-6167-406c-bc4e-94a82e072bec.png">

**Comment by alixander** [comment]
> It was much better when I divided both of those values by 2.

@pangloss is this just an arbitrary zoom level to get text readable though? which differs depending on the diagram? i'm not sure what D2 can do here

**Comment by pangloss** [comment]
I mean text readability. The default text/label size is constant, no matter how many nodes in the diagram. At some point it's personal preference, but I am aiming for something about the size I pasted above, which I think balances readability with maximizing the amount of diagram visible.

**Comment by alixander** [comment]
oh i think i know what you're talking about. you're saying, scale such that a "font-size: 12" word always takes up the same amount of pixels on the screen.

is that right?

**Comment by pangloss** [comment]
Yes

**Comment by pangloss** [comment]
Hey, this one seems to have fallen by the wayside. Is there still any interest in resolving it?

**Comment by alixander** [comment]
thanks for bumping this. @gavin-ts any ideas how to do this?

**Comment by gavin-ts** [comment]
I think if we can just merge in #721, that combined with zooming to 50% in the browser should have the desired effect.

**Comment by alixander** [comment]
i don't think that's the desired effect.

"scale such that a "font-size: 12" word always takes up the same amount of pixels on the screen."

i think optimizing for text readability has a wide range of zoom levels depending on size of diagram.

**Comment by gavin-ts** [comment]
> "scale such that a "font-size: 12" word always takes up the same amount of pixels on the screen."

when we set the dimensions, font-size: 12 is always the same size

**Comment by bo-ku-ra** [comment]
was this not implemented in d2 v0.5.0? (https://github.com/terrastruct/d2/pull/721)

**Comment by gavin-ts** [comment]
> was this not implemented in d2 v0.5.0? (#721)

#1346 is still waiting on review
